File: Arrays/2dArray/2dArray.cpp
#include <iostream>
#include <fstream>
#include <string.h>
#include <sstream>
#include <vector> 

using namespace std;


void print_2d_vector(vector<vector<int> > arr_in) {
	for(size_t i = 0; i < arr_in.size();i++) {
		for(size_t j = 0; j < arr_in[i].size();j++)	{
			cout << arr_in[i][j];
		}
		cout << endl;
	}

}
class Array2d {
	private:
	vector<vector<int> > arr;
	public:
	Array2d(vector<vector<int> > arr_in) {
		arr = arr_in;
		//print_2d_vector(arr);
	}

	int solve() {
		int max = -360;
		for (size_t i = 1; i<arr.size()-1;i++) {
			for (size_t j = 1; j < arr[0].size()-1 ; j++) {
				int sum = sumI(i,j);
				if (sum > max) {
					max = sum;
				}

			}
		}
		return max;
	}

	int sumI(int x, int y)
	{
		// Assumes valid I shape
		
		//int ret = arr[x-1][y-1]+arr[x][y-1]+arr[x+1][y-1]+arr[x][y]+arr[x-1][y+1]+arr[x][y+1]+arr[x+1][y+1];
		int ret = arr[x-1][y-1]+arr[x-1][y]+arr[x-1][y+1]+arr[x][y]+arr[x+1][y-1]+arr[x+1][y]+arr[x+1][y+1];

		//cout << x << " " << y << " "<< ret<< endl;

		return ret;
	}
};

int main() {
	//std::cout << "Hello world!" << std::endl;
	fstream myfile;
	myfile.open("input2.txt");
	string line,row;
	vector<vector<int> > arr(6);
	if(myfile.is_open()) {
		int i= 0;
		while(getline(myfile,row, '\n')) {

			stringstream ssin(row);
			string word;
			int j = 0;
			arr[i].resize(6);
			while(ssin.good()) {
				ssin >> word;
				arr[i][j] = atoi(word.c_str());
				j++;
			}					
			i++;
		}
	}
	myfile.close();

	Array2d * sol = new Array2d(arr);
	cout << sol->solve() << endl;

}

File: Arrays/2dArray/2dArray.py
# 2dArray.py
# https://www.hackerrank.com/challenges/2d-array/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays 
import math

def main():
	# import from input
	f = open("input.txt", "r")
	arr = []
	for i in f:
		#print([int(i) for i in i.rstrip().split()])
		#arr.append(list())
		arr.append([int(i) for i in i.rstrip().split()])
	f.close()
	#print(arr)
	sol = array2d(arr).solve()
	print(sol)
	#invoke class

class array2d:
	
	def __init__(self, arr_in):
		self.arr = arr_in

	def solve(self):
		max_sum = float("-inf")
		for x in range(1,len(self.arr[0])-1):
			for y in range(1,len(self.arr[0])-1):
				sum_pos = self.sumI(x,y)
				if sum_pos > max_sum:
					max_sum = sum_pos

		return max_sum
	def sumI(self,x, y):
		#Assumes hour glass is valid
		# Sums top row

		sum_hourglass = sum(self.arr[x-1][y-1:y+2])
		sum_hourglass += self.arr[x][y]
		sum_hourglass += sum(self.arr[x+1][y-1:y+2])
		return sum_hourglass

main()